server/src/main.rs

```diff
@@ -1,3 +1,153 @@
+use std::io::{ErrorKind, Read, Write};
+use std::net::TcpListener;
+use std::sync::mpsc::channel;
+use std::thread::{sleep, spawn};
+use std::time::Duration;
+
+use http::header::{CONNECTION, SEC_WEBSOCKET_ACCEPT, UPGRADE};
+use http::{HeaderMap, Response, Version};
+use httparse::{Request, EMPTY_HEADER};
+use sha1_smol::Sha1;
+
 fn main() {
-    println!("Hello, world!");
+    let server = TcpListener::bind("127.0.0.1:8080").unwrap();
+    server.set_nonblocking(true).unwrap();
+
+    let (sender, receiver) = channel();
+
+    let mut clients = vec![];
+
+    loop {
```

Comment:
there must be a better way than infintely looping and waiting for server connections some kind of a event handler some reactive framework etc.

Reply:
The framework will probably do exactly the same. But if there is a good framework, we might try it.